languages/rust/src/host/mod.rs

```diff
@@ -0,0 +1,476 @@
+                .cast_to_instance(i),
+            Value::Number(Numeric::Float(f)) => host
+                .get_class_from_type::<f64>()
+                .unwrap()
+                .cast_to_instance(f),
+            Value::List(v) => host
+                .get_class_from_type::<Vec<Term>>()
+                .unwrap()
+                .cast_to_instance(v),
+            Value::String(s) => host
+                .get_class_from_type::<String>()
+                .unwrap()
+                .cast_to_instance(s),
+            Value::Dictionary(d) => host
+                .get_class_from_type::<HashMap<Name, Term>>()
+                .unwrap()
+                .cast_to_instance(d.fields),
+            Value::ExternalInstance(ExternalInstance { instance_id, .. }) => host
+                .get_instance(instance_id)
+                .expect("instance not found")
+                .clone(),
+            v => {
+                tracing::warn!(value = ?v, "invalid conversion attempted");
+                return None;
+            }
+        };
+        Some(instance)
+    }
+}
+
+impl FromPolar for () {
+    fn from_polar(term: &Term, _host: &mut Host) -> Option<Self> {
+        if let Value::List(l) = term.value() {
+            if l.is_empty() {
+                Some(())
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    fn from_polar_list(terms: &[Term], _host: &mut Host) -> Option<Self> {
+        if terms.is_empty() {
+            Some(())
+        } else {
+            None
+        }
+    }
+}
+
+impl<A> FromPolar for (A,)
+where
+    A: FromPolar,
+{
+    fn from_polar(_term: &Term, _host: &mut Host) -> Option<Self> {
+        None
+    }
+
+    fn from_polar_list(terms: &[Term], host: &mut Host) -> Option<Self> {
+        if terms.len() == 1 {
+            A::from_polar(&terms[0], host).map(|a| (a,))
+        } else {
+            None
+        }
+    }
+}
+
+impl<A, B> FromPolar for (A, B)
+where
+    A: FromPolar,
+    B: FromPolar,
+{
+    fn from_polar(_term: &Term, _host: &mut Host) -> Option<Self> {
+        None
+    }
+
+    fn from_polar_list(terms: &[Term], host: &mut Host) -> Option<Self> {
+        if terms.len() == 2 {
+            let a = A::from_polar(&terms[0], host)?;
+            let b = B::from_polar(&terms[1], host)?;
+            Some((a, b))
+        } else {
+            None
+        }
+    }
+}
+
+/// Marker trait: implements "ToPolar" via a registered class
+pub trait HostClass {}
+
+impl<C: 'static + Clone + HostClass> FromPolar for C {
+    fn from_polar(term: &Term, host: &mut Host) -> Option<Self> {
+        match term.value() {
+            Value::ExternalInstance(ExternalInstance { instance_id, .. }) => {
+                let instance = host.get_instance(*instance_id)?;
+                instance.instance.downcast_ref::<C>().cloned()
+            }
+            _ => None,
+        }
+    }
+}
+
+impl<C: 'static + Clone + HostClass> ToPolar for C {
```

Comment:
Are there any common use cases that would be eliminated by requiring this to be `'static`?

Reply:
Types with references?

We can try and relax this and add lifetimes to `Polar` instances so that we know that any cached value doesn't exceed the lifetime of the polar instance.

Reply:
w/r/t @samscott89's comment, it would also be cool to reduce the amount of cloning we're currently doing in this implementation.